opm/autodiff/WellInterface_impl.hpp

```diff
@@ -863,7 +900,76 @@ namespace Opm
 
         for (int p = 0; p < np; ++p) {
             well_state.wellReservoirRates()[well_rate_index + p] = voidage_rates[p];
+        }    
+    }
+
+    template<typename TypeTag>
+    void
+    WellInterface<TypeTag>::closeWellsAndCompletions(WellTestState& wellTestState)
+    {
+        if (wellTestState.hasWell(name(), WellTestConfig::Reason::ECONOMIC)) {
+            assert(!well_ecl_->getAutomaticShutIn());
+            well_controls_stop_well(wellControls());
+        }
+
+        const auto& connections = well_ecl_->getConnections(current_step_);
+        int perfIdx = 0;
+        for (const auto& connection : connections) {
+            if (wellTestState.hasCompletion(name(), connection.complnum)) {
+                well_index_[perfIdx] = 0.0;
+            }
+            perfIdx++;
+        }
+    }
+
+    template<typename TypeTag>
+    SimulatorReport
+    WellInterface<TypeTag>::solveWellEq(Simulator& ebosSimulator, WellState& well_state, const double dt, const std::vector<double>& B_avg, bool terminal_output)
+    {
+        const int max_iter = param_.max_welleq_iter_;
+        int it  = 0;
+        bool converged;
+        WellState well_state0 = well_state;
+        do {
+            assembleWellEq(ebosSimulator, dt, well_state, true);
+
+            ConvergenceReport report;
+            report = getWellConvergence(B_avg);
+            converged = report.converged;
+
+            if (converged) {
+                break;
+            }
+
+            ++it;
+            solveEqAndUpdateWellState(well_state);
+
+            wellhelpers::WellSwitchingLogger logger;
+            updateWellControl(well_state, logger);
+            initPrimaryVariablesEvaluation();
+        } while (it < max_iter);
+
+        if (converged) {
+            if ( terminal_output ) {
```

Comment:
Since this function is for individual well, it will be helpful to indicate the well name. I am not totally sure they are only used for wellTesting yet. If yes, the message should indicate that it is for well testing running.

Reply:
I hope this can replace the solveWellEq in BlackoilWellModel one day. Currently this can not be done due to the way the group control is implemented. 

I will add the well name in the output.

Reply:
And also, it is possible that even in the future, solveWellEq can be difficult to happen on the individual well basis when group or network is involved, which means the interaction between wells happen.